homeassistant/components/evohome/__init__.py

```diff
@@ -44,204 +45,240 @@
+                access_token_expires
+            )
 
-        self._available = False  # should become True after first update()
+        except (requests.exceptions.RequestException,
+                evohomeclient2.AuthenticationError) as err:
+            if not _handle_exception(err):
+                return False
 
-    @callback
-    def _connect(self, packet):
-        if packet['to'] & self._type and packet['signal'] == 'refresh':
-            self.async_schedule_update_ha_state(force_refresh=True)
+        else:
+            if access_token != self.client.access_token:
+                await self._save_auth_tokens()
 
-    def _handle_exception(self, err):
+        finally:
+            self.params[CONF_PASSWORD] = 'REDACTED'
+
+        loc_idx = self.params[CONF_LOCATION_IDX]
         try:
-            raise err
+            self.config = client.installation_info[loc_idx][GWS][0][TCS][0]
 
-        except evohomeclient2.AuthenticationError:
+        except IndexError:
             _LOGGER.error(
-                "Failed to (re)authenticate with the vendor's server. "
-                "This may be a temporary error. Message is: %s",
-                err
+                "Config error: '%s' = %s, but its valid range is 0-%s. "
+                "Unable to continue. "
+                "Fix any configuration errors and restart HA.",
+                CONF_LOCATION_IDX, loc_idx, len(client.installation_info) - 1
             )
+            return False
 
-        except requests.exceptions.ConnectionError:
-            # this appears to be common with Honeywell's servers
-            _LOGGER.warning(
-                "Unable to connect with the vendor's server. "
-                "Check your network and the vendor's status page."
-            )
+        else:
+            self.tcs = \
+                client.locations[loc_idx]._gateways[0]._control_systems[0]  # noqa: E501; pylint: disable=protected-access
+
+        _LOGGER.debug("Config = %s", self.config)
+
+        return True
 
-        except requests.exceptions.HTTPError:
-            if err.response.status_code == HTTP_SERVICE_UNAVAILABLE:
-                _LOGGER.warning(
-                    "Vendor says their server is currently unavailable. "
-                    "This may be temporary; check the vendor's status page."
-                )
-
-            elif err.response.status_code == HTTP_TOO_MANY_REQUESTS:
-                _LOGGER.warning(
-                    "The vendor's API rate limit has been exceeded. "
-                    "So will cease polling, and will resume after %s seconds.",
-                    (self._params[CONF_SCAN_INTERVAL] * 3).total_seconds()
-                )
-                self._timers['statusUpdated'] = datetime.now() + \
-                    self._params[CONF_SCAN_INTERVAL] * 3
+    async def _load_auth_tokens(self) -> Tuple[str, str, datetime]:
+        store = self.hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
```

Comment:
This is literally why config entries exist, to manage auth tokens. We should probably just migrate directly to that .

Reply:
OMG - are you talking about that config flow thing? You couldn't sign-post me to an integration that is already doing it?

Reply:
There are 60 config flows to choose from 😉 

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/generated/config_flows.py

An example is https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/met/config_flow.py#L166 and there is a ton of docs about it on https://developers.home-assistant.io/

Reply:
I just saw your post re: `climate-1.0`.

I am able to refactor 4 climate entities (**evohome**, **geniushub**, **honeywell**, **incomfort**), and have access to these systems fro testing. I have even submitted a PR for two of them.  

However, **evohome**, and **geniushub** wont be easy, and so I am waiting for feedback on the two I've already done (**honeywell**, **incomfort**), which haven't been merged, before I start work on them.

There is the added complicated of deprecating EU out of **honeywell**, and that obligates `water_heater` support in **evohome** (one of the two purposes of this PR)...

Additionally, the API was recently changed by the vendor, and rate limiting is on initial OAuth rather than ongoing polling - this means development/testing is virtually impossible unless access tokens can be cached/restored (the other purpose of this PR).

So...

Can I get this PR accepted as is, using storage, I'll then have time to work on `climate-1.0`, and after that, I'll go back and migrate to config entries?

Or is it all too late?

Otherwise, can I put all this code into `climate-1.0`? - will mean a large PR

Reply:
We can migrate later from storage.

Reply:
I'm also fine to take this as a big PR into climate 1.0.

Reply:
Great! Makes my life so much easier! I'll see if I can have this ready for review in <24h.